lib/libc/mips/string/jdw57_memcpy.c

```diff
@@ -0,0 +1,281 @@
+			/*
+			 * Try to align operands.  This cannot be done
+			 * unless the low bits match.
+			 */
+			if ((t ^ (vaddr_t)dst) & wmask || length < wsize)
+				t = length;
+			else
+				t = wsize - (t & wmask);
+			length -= t;
+			dst += t;
+			src += t;
+			t = -t;
+			MIPSLOOP(t, -1, dst[t]=src[t];, 1);
+		}
+		/*
+		 * If pointers are bigger than words, try to copy by words.
+		 */
+		if (bigptr) {
+			t = (vaddr_t)src;	/* only need low bits */
+			if ((t | (vaddr_t)dst) & pmask) { // XXX make sure dst cast gets virtual address
+				/*
+				 * Try to align operands.  This cannot be done
+				 * unless the low bits match.
+				 */
+				if ((t ^ (vaddr_t)dst) & pmask || length < psize)
+					t = length / wsize;
+				else
+					t = (psize - (t & pmask)) / wsize;
+				if (t) {
+					length -= t*wsize;
+					dst += t*wsize;
+					src += t*wsize;
+					t = -t*wsize;
+					MIPSLOOP(t, -8, *((word * CAPABILITY)(dst+t)) = *((word * CAPABILITY)(src+t));, 8/*wsize*/);
+				}
+			}
+		}
+		/*
+		 * Copy whole words, then mop up any trailing bytes.
+		 */
+		t = length / psize;
+		if (t) {
+			src += t*psize;
+			dst += t*psize;
+			t = -(t*psize);
+			//if (t) MIPSLOOP(t, -psize, *((ptr * CAPABILITY)(dst+t)) = *((ptr * CAPABILITY)(src+t));, PWIDTH);
+			#if !defined(_MIPS_SZCAP)
+		    MIPSLOOP(t, -psize, *((ptr * CAPABILITY)(dst+t)) = *((ptr * CAPABILITY)(src+t));, 8);
+			#elif _MIPS_SZCAP==128
+				MIPSLOOP(t, -psize, *((ptr * CAPABILITY)(dst+t)) = *((ptr * CAPABILITY)(src+t));, 16);
+			#elif _MIPS_SZCAP==256
+				MIPSLOOP(t, -psize, *((ptr * CAPABILITY)(dst+t)) = *((ptr * CAPABILITY)(src+t));, 32);
+			#endif
+		}
+		t = length & pmask;
+		if (t) {
+			src += t;
+			dst += t;
+			t = -t;
+			if (t) MIPSLOOP(t, -1, dst[t]=src[t];, 1);
+		}
+	}	else {
+		/*
+		 * Copy backwards.  Otherwise essentially the same.
+		 * Alignment works as before, except that it takes
+		 * (t&wmask) bytes to align, not wsize-(t&wmask).
+		 */
+		src += length;
+		dst += length;
+		t = (vaddr_t)src;
+		if ((t | (vaddr_t)dst) & wmask) {
+			if ((t ^ (vaddr_t)dst) & wmask || length <= wsize)
+				t = length;
+			else
+				t &= wmask;
+			length -= t;
+			dst -= t;
+			src -= t;
+			t--;
+			MIPSLOOP(t, 0, dst[t]=src[t];, -1);
+		}
+		if (bigptr) {
+			t = (vaddr_t)src;	/* only need low bits */
+			if ((t | (vaddr_t)dst) & pmask) {
+				if ((t ^ (vaddr_t)dst) & pmask || length < psize)
+					t = length / wsize;
+				else
+					t = (psize - (t & pmask)) / wsize;
+				if (t) {
+					length -= t*wsize;
+					dst -= t*wsize;
+					src -= t*wsize;
+					t = ((t-1)*wsize);
+					MIPSLOOP(t, 0, *((word * CAPABILITY)(dst+t)) = *((word * CAPABILITY)(src+t));, -8/*wsize*/);
+				}
+			}
+		}
+		t = length / psize;
+		if (t) {
+			src -= t*psize;
+			dst -= t*psize;
+			t = ((t-1)*psize);
+#if !defined(_MIPS_SZCAP)
+		  MIPSLOOP(t, 0, *((ptr * CAPABILITY)(dst+t)) = *((ptr * CAPABILITY)(src+t));, -8/*sizeof(ptr)*/);
```

Comment:
Indentation seems wrong (too few tabs and too many spaces.) Also wrap the MIPSLOOP lines at 80 columns.